src/sdk/openmldb_api.cc

```diff
@@ -0,0 +1,378 @@
+    return *this;
+}
+
+RequestRow& RequestRow::operator<<(const int64_t value) {
+    parameter_types->AddColumnType(::hybridse::sdk::kTypeInt64);
+    record.push_back(value);
+    return *this;
+}
+
+RequestRow& RequestRow::operator<<(const TimeStamp value) {
+    parameter_types->AddColumnType(::hybridse::sdk::kTypeTimestamp);
+    record.push_back(value.get_Timestamp());
+    return *this;
+}
+
+RequestRow& RequestRow::operator<<(const Date value) {
+    parameter_types->AddColumnType(::hybridse::sdk::kTypeDate);
+    record.push_back(value.get_year());
+    record.push_back(value.get_month());
+    record.push_back(value.get_day());
+    return *this;
+}
+
+RequestRow& RequestRow::operator<<(const float value) {
+    parameter_types->AddColumnType(::hybridse::sdk::kTypeFloat);
+    record.push_back(value);
+    return *this;
+}
+
+RequestRow& RequestRow::operator<<(const double value) {
+    parameter_types->AddColumnType(::hybridse::sdk::kTypeDouble);
+    record.push_back(value);
+    return *this;
+}
+
+RequestRow& RequestRow::operator<<(const std::string value) {
+    parameter_types->AddColumnType(::hybridse::sdk::kTypeString);
+    str_length += value.length();
+    record.push_back(value);
+    return *this;
+}
+
+RequestRow& RequestRow::operator<<(const char* value) {
+    parameter_types->AddColumnType(::hybridse::sdk::kTypeString);
+    str_length += strlen(value);
+    record.push_back((std::string)value);
+    return *this;
+}
+
+RequestRow& RequestRow::operator<<(const OpenmldbNull value) {
+    parameter_types->AddColumnType(::hybridse::sdk::kTypeUnknow);
+    record.push_back(value);
+    return *this;
+}
+
+void RequestRow::reset() {
+    record.clear();
+    parameter_types = std::make_shared<hybridse::sdk::ColumnTypes>();
+    str_length = 0;
+}
+
+bool execute(const OpenmldbHandler& handler, const std::string& sql) {
+    auto rs = (handler.get_router())->ExecuteSQL(sql, handler.get_status());
+    if (rs != NULL) resultset_last = rs;
+    return (handler.get_status())->IsOK();
+}
+
+bool execute(const OpenmldbHandler& handler, const std::string& db, const std::string& sql) {
+    auto rs = (handler.get_router())->ExecuteSQL(db, sql, handler.get_status());
+    if (rs != NULL) resultset_last = rs;
+    return (handler.get_status())->IsOK();
+}
+
+bool execute_parameterized(const OpenmldbHandler& handler, const std::string& db, const std::string& sql,
+                           const ParameterRow& para) {
+    auto pr = para.get_parameter_row();
+    auto rs = (handler.get_router())->ExecuteSQLParameterized(db, sql, pr, handler.get_status());
+    if (rs != NULL) resultset_last = rs;
+    return (handler.get_status())->IsOK();
+}
+
+bool execute_request(const RequestRow& req) {
+    auto rr = req.get_request_row();
+    auto rs = ((req.get_handler())->get_router())
+                  ->ExecuteSQLRequest(req.get_db(), req.get_sql(), rr, (req.get_handler())->get_status());
+    if (rs != NULL) resultset_last = rs;
```

Comment:
how to guarantee thread safe

Reply:
This style `resultset_last`, can't do thread safe, if 2 threads both set it, thread 1 may can't get the right rs, cuz thread 2 overwrite it.

Reply:
If we need it, let's fix it later